File: tests/solutions_js/longest_increasing_subsequence.js
class Solution {
    lengthOfLIS(nums) {
        const tails = [];
        for (const num of nums) {
            let lo = 0, hi = tails.length;
            while (lo < hi) {
                const mid = (lo + hi) >> 1;
                if (tails[mid] < num) lo = mid + 1;
                else hi = mid;
            }
            if (lo === tails.length) {
                tails.push(num);
            } else {
                tails[lo] = num;
            }
        }
        return tails.length;
    }
}

File: tests/solutions_cpp/longest_increasing_subsequence.cpp
#include "json_helper.h"

class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        vector<int> tails;
        for (int n : nums) {
            auto it = lower_bound(tails.begin(), tails.end(), n);
            if (it == tails.end()) tails.push_back(n);
            else *it = n;
        }
        return (int)tails.size();
    }
};

int main() {
    auto args = parse_args();
    Solution sol;
    auto nums = to_vector_int(args[0]);
    int result = sol.lengthOfLIS(nums);
    cout << result << endl;
    return 0;
}
